src/pkgcheck/checks/codingstyle.py

```diff
@@ -903,3 +903,102 @@ def feed(self, pkg):
                             lineno=lineno - 1, pkg=pkg)
+
+
+class EbuildUnquotedVariable(results.LineResult, results.Style):
+    """Variable is unquoted, but should be quoted"""
+
+    def __init__(self, variable, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+
+    @property
+    def desc(self):
+        return f"{self.lineno}: Variable {self.variable} should be quoted"
+
+
+class EclassUnquotedVariable(results.EclassResult, results.Style):
+    """Variable is unquoted, but should be quoted"""
+
+    def __init__(self, variable, lineno, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+        self.lineno = lineno
+
+    @property
+    def desc(self):
+        return f"{self.lineno}: Variable {self.variable} should be quoted"
+
+
+class _UnquotedVariablesCheck(Check):
+    message_commands = (
+        "die", "echo", "eerror", "einfo", "elog", "eqawarn", "ewarn"
+    )
+    var_names = (
+        "D", "DISTDIR", "FILESDIR", "S", "T", "ROOT", "BROOT", "WORKDIR", "ED",
+        "EPREFIX", "EROOT", "SYSROOT", "ESYSROOT", "TMPDIR", "HOME",
+        # variables for multibuild.eclass
+        "BUILD_DIR",
+    )
+
+    def _var_needs_quotes(self, pkg, node):
+        pnode = node.parent
+        while pnode != node:
+            if pnode.type == 'string':
+                return False
+            elif pnode.type == 'variable_assignment':
+                # Variable is part of a shell assignment, and does not need to
+                # be quoted. For example S=${WORKDIR}/${PN} is OK.
+                return False
+            elif pnode.type == 'test_command':
+                # Variable sits inside a [[ ]] test command and it's OK not to
+                # be quoted
+                return False
+            elif pnode.type == 'heredoc_body':
+                # Variable is being used in a heredoc body, no need to specify
+                # quotes.
+                return False
+            elif pnode.type == 'command':
+                cmd = pkg.node_str(pnode.child_by_field_name('name'))
+                return cmd not in self.message_commands
+            elif pnode.type in 'array':
+                # Variable is sitting unquoted in an array
+                return True
+            pnode = pnode.parent
+        return True
+
+    def _feed(self, pkg):
+        if pkg.tree.root_node.has_error:
+            # Do not run this check if the parse tree contains errors, as it
+            # might result in false positives. This check appears to be quite
+            # expensive though...
+            return
+        for var_node, _ in bash.var_query.captures(pkg.tree.root_node):
+            var_name = pkg.node_str(var_node)
+            if var_name in self.var_names:
+                if self._var_needs_quotes(pkg, var_node):
+                    lineno, _ = var_node.start_point
+                    yield var_name, lineno+1
```

Comment:
Same as mentioned above, I'd push results into a dictionary here and then iterate over them after AST walking is done, returning a result per variable per ebuild/eclass with multiple lines reported as necessary.